src/cpu/Dasm.cc

```diff
@@ -23,82 +22,90 @@ void appendAddrAsHex(std::string& output, uint16_t addr)
-		switch (buf[0]) {
+		switch (bin[0]) {
 			case 0xCB:
-				buf[1] = interface.peekMem(pc + 1, time);
-				return {mnemonic_cb[buf[1]], 2};
+				return {mnemonic_cb[bin[1]], 2};
 			case 0xED:
-				buf[1] = interface.peekMem(pc + 1, time);
-				return {mnemonic_ed[buf[1]], 2};
+				return {mnemonic_ed[bin[1]], 2};
 			case 0xDD:
 			case 0xFD:
-				r = (buf[0] == 0xDD) ? "ix" : "iy";
-				buf[1] = interface.peekMem(pc + 1, time);
-				if (buf[1] != 0xcb) {
-					return {mnemonic_xx[buf[1]], 2};
+				r = (bin[0] == 0xDD) ? "ix" : "iy";
+				if (bin[1] != 0xCB) {
+					return {mnemonic_xx[bin[1]], 2};
 				} else {
-					buf[2] = interface.peekMem(pc + 2, time);
-					buf[3] = interface.peekMem(pc + 3, time);
-					return {mnemonic_xx_cb[buf[3]], 4};
+					return {mnemonic_xx_cb[bin[3]], 4};
 				}
 			default:
-				return {mnemonic_main[buf[0]], 1};
+				return {mnemonic_main[bin[0]], 1};
 		}
 	}();
 
 	for (int j = 0; s[j]; ++j) {
 		switch (s[j]) {
 		case 'B':
-			buf[i] = interface.peekMem(narrow_cast<uint16_t>(pc + i), time);
 			strAppend(dest, '#', hex_string<2>(
-				static_cast<uint16_t>(buf[i])));
+				static_cast<uint16_t>(bin[i])));
 			i += 1;
 			break;
 		case 'R':
-			buf[i] = interface.peekMem(narrow_cast<uint16_t>(pc + i), time);
-			appendAddr(dest, uint16_t(pc + 2 + static_cast<int8_t>(buf[i])));
+			appendAddr(dest, uint16_t(pc + 2 + static_cast<int8_t>(bin[i])));
 			i += 1;
 			break;
 		case 'A':
 		case 'W':
-			buf[i + 0] = interface.peekMem(narrow_cast<uint16_t>(pc + i + 0), time);
-			buf[i + 1] = interface.peekMem(narrow_cast<uint16_t>(pc + i + 1), time);
-			appendAddr(dest, buf[i] + buf[i + 1] * 256);
+			appendAddr(dest, bin[i] + bin[i + 1] * 256);
 			i += 2;
 			break;
 		case 'X':
-			buf[i] = interface.peekMem(narrow_cast<uint16_t>(pc + i), time);
-			strAppend(dest, '(', r, sign(buf[i]), '#',
-			     hex_string<2>(abs(buf[i])), ')');
+			strAppend(dest, '(', r, sign(bin[i]), '#',
+			     hex_string<2>(abs(bin[i])), ')');
 			i += 1;
 			break;
 		case 'Y':
-			strAppend(dest, r, sign(buf[2]), '#', hex_string<2>(abs(buf[2])));
+			strAppend(dest, r, sign(bin[2]), '#', hex_string<2>(abs(bin[2])));
 			break;
 		case 'I':
 			dest += r;
 			break;
-		case '!':
-			dest = strCat("db     #ED,#", hex_string<2>(buf[1]),
-			              "     ");
-			return 2;
-		case '@':
-			dest = strCat("db     #", hex_string<2>(buf[0]),
+		case '!': // used when bin contains invalid z80 instruction
+			dest = strCat("db     #ED         ");
+			// skip assertion at the end of function
+			return 1;
+		case '@': // used when bin contains invalid z80 instruction
+			dest = strCat("db     #", hex_string<2>(bin[0]),
 			              "         ");
+			// skip assertion at the end of function
+			return 1;
+		case '#': // used when bin contains invalid z80 instruction
+			dest = strCat("db     #", hex_string<2>(bin[0]),
+			              "         ");
+			// skip assertion at the end of function
 			return 1;
```

Comment:
I made them return 0 in case of invalid Z80 instruction, otherwise the comparison with `instructionLength(std::span<const uint8_t> bin)` will fail in unittest.

Reply:
These should return the length of the invalid instruction, not 0. Or actually didn't we already change the return type of dasm() to void?

From the disassembly point of view, there are no invalid instructions. For example the sequence `DD 00` is an instruction with length 2 (not 0) that produces the string `db #DD,#00`. And similar for all other _invalid_ instructions. And this is also how the real Z80 sees this sequence: it's a 2-byte instruction with the same effect as `nop`.

So these `return` statements in `case '#' '!' '@'` should be removed. And the statement above should be adjusted to print all `i` bytes.

The unittest should not do anything special for _invalid_ instructions. Again, from the disassembly point of view there are no invalid instructions.